windows/devices/uwb/UwbConnector.cxx

```diff
@@ -915,8 +939,90 @@ UwbConnector::CallbacksPresent()
 }
 
 void
-UwbConnector::DeregisterEventCallback(::uwb::RegisteredCallbackToken* token)
+UwbConnector::DeregisterEventCallback(std::weak_ptr<::uwb::RegisteredCallbackToken> token)
 {
-    // TODO implement
+    auto tok = token.lock();
+    if (not tok) {
+        return;
+    }
     std::lock_guard eventCallbacksLockExclusive{ m_eventCallbacksGate };
+
+    DeregisterEventCallback(*tok);
 }
+
+void
+UwbConnector::DeregisterEventCallback(::uwb::RegisteredCallbackToken token)
+{
+    auto callbackId = token.callbackId;
+    auto isDeviceEventCallback = token.isDeviceEventCallback;
+
+    // first remove the token from the list of tokens
+    for (auto it = std::cbegin(m_tokens); it != std::cend(m_tokens); it++) {
+        if ((*it)->callbackId == callbackId) {
+            m_tokens.erase(it);
+            break;
+        }
+    }
+
+    // deal with the id map and list of callbacks
+    if (isDeviceEventCallback) {
+        // first remove the id from the id map
+        auto result = m_deviceEventCallbacksIdMap.find(callbackId);
+        if (result == std::cend(m_deviceEventCallbacksIdMap)) {
+            return;
+        }
+        m_deviceEventCallbacksIdMap.erase(result);
+
+        // now remove the id from the list of callbacks
+        for (auto it = std::cbegin(m_deviceEventCallbacks); it != std::cend(m_deviceEventCallbacks); it++) {
+            if (*it == callbackId) {
+                m_deviceEventCallbacks.erase(it);
+                return;
+            }
+        }
+    } else {
+        // first remove the id from the id map
+        auto result = m_sessionEventCallbacksIdMap.find(callbackId);
+        if (result == std::cend(m_sessionEventCallbacksIdMap)) {
+            return;
+        }
+        m_sessionEventCallbacksIdMap.erase(result);
+
+        // now remove the id from the list of callbacks
+        for (auto sessionIt = std::cbegin(m_sessionEventCallbacks); sessionIt != std::cend(m_sessionEventCallbacks); sessionIt++) {
+            auto [sessionId, sessionCallbackIds] = *sessionIt;

```

Comment:
You could use the structured binding directly in the for loop:

`for (const auto& [sessionId, sessionCallbackIds] : m_sessionEventCallbacks)`

Reply:
the for loop also erases sessionIt from m_sessionEventCallbacks if it becomes empty, so this won't be possible